drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,684 @@
+/*
+ * Copyright (c) 2021 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_i2c
+
+#include <errno.h>
+#include <kernel.h>
+#include <devicetree.h>
+#include <drivers/pinctrl.h>
+#include <drivers/i2c.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_gd32, CONFIG_I2C_LOG_LEVEL);
+
+#include "i2c-priv.h"
+
+struct i2c_gd32_config {
+	uint32_t reg;
+	uint32_t bitrate;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+	void (*irq_cfg_func)(void);
+};
+
+struct i2c_gd32_data {
+	struct k_sem bus_mutex;
+	struct k_sem sync_sem;
+	uint32_t dev_config;
+	uint16_t addr1;
+	uint16_t addr2;
+	uint32_t xfer_len;
+	struct i2c_msg *current;
+	uint8_t is_berr;
+	uint8_t is_lostarb;
+	uint8_t is_aerr;
+	uint8_t is_restart;
+	uint8_t is_stopped;
+};
+
+static inline void i2c_gd32_enable(const struct i2c_gd32_config *cfg)
+{
+	I2C_CTL0(cfg->reg) |= I2C_CTL0_I2CEN;
+}
+
+static inline void i2c_gd32_disable(const struct i2c_gd32_config *cfg)
+{
+	I2C_CTL0(cfg->reg) &= ~I2C_CTL0_I2CEN;
+}
+
+static inline void i2c_gd32_enable_interrupts(const struct i2c_gd32_config *cfg)
+{
+	I2C_CTL1(cfg->reg) |= I2C_CTL1_ERRIE;
+	I2C_CTL1(cfg->reg) |= I2C_CTL1_EVIE;
+	I2C_CTL1(cfg->reg) |= I2C_CTL1_BUFIE;
+}
+
+static inline void i2c_gd32_disable_interrupts(const struct i2c_gd32_config *cfg)
+{
+	I2C_CTL1(cfg->reg) &= ~I2C_CTL1_ERRIE;
+	I2C_CTL1(cfg->reg) &= ~I2C_CTL1_EVIE;
+	I2C_CTL1(cfg->reg) &= ~I2C_CTL1_BUFIE;
+}
+
+static inline void i2c_gd32_xfer_read(struct i2c_gd32_data *data,
+				      const struct i2c_gd32_config *cfg)
+{
+	data->current->len--;
+	*data->current->buf = I2C_DATA(cfg->reg);
+	data->current->buf++;
+
+	if (data->xfer_len && (data->current->len == 0U)) {
+		data->current++;
+	}
+}
+
+static inline void i2c_gd32_xfer_write(struct i2c_gd32_data *data,
+				       const struct i2c_gd32_config *cfg)
+{
+	data->current->len--;
+	I2C_DATA(cfg->reg) = *data->current->buf;
+	data->current->buf++;
+
+	if (data->xfer_len && (data->current->len == 0U)) {
+		data->current++;
+	}
+}
+
+static void i2c_gd32_handle_rbne(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->xfer_len) {
+		switch (data->xfer_len) {
+		case 1:
+			/* Reception for N == 1. */
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+
+			k_sem_give(&data->sync_sem);
+
+			break;
+		case 2:
+			__fallthrough;
+		case 3:
+			/*
+			 * Reception for N == 2 or
+			 * reception for N > 3 when xfer_len == 3.
+			 */
+			I2C_CTL1(cfg->reg) &= ~I2C_CTL1_BUFIE;
+			break;
+		default:
+			/* Reception for N > 3 when xfer_len > 3. */
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+			break;
+		}
+	} else {
+		if ((data->current->flags & I2C_MSG_STOP) &&
+		    (data->is_stopped == 0U)) {
+			data->is_stopped = 1U;
+			I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+		}
+
+		k_sem_give(&data->sync_sem);
+	}
```

Comment:
suggestion: treat `0` case as another case

Reply:
As I know ``else`` are dead code, reception process will not come into here.

Reply:
Move the ``0`` case into switch and add a comment to ignore it.